network/endpoint_linux.go

```diff
@@ -53,205 +54,200 @@ func (nw *network) newEndpointImpl(
+						nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
+					}
+					epClient = NewTransparentVlanEndpointClient(nw, epInfo, hostIfName, contIfName, vlanid, localIP, nl, plc)
+				} else {
+					logger.Info("OVS client")
+					if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
+						nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
+					}
+
+					epClient = NewOVSEndpointClient(
+						nw,
+						epInfo,
+						hostIfName,
+						contIfName,
+						vlanid,
+						localIP,
+						nl,
+						ovsctl.NewOvsctl(),
+						plc)
 				}
-				epClient = NewTransparentVlanEndpointClient(nw, epInfo, hostIfName, contIfName, vlanid, localIP, nl, plc)
+			} else if nw.Mode != opModeTransparent {
+				logger.Info("Bridge client")
+				epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
+			} else if epInfo.NICType == cns.DelegatedVMNIC {
+				logger.Info("Secondary client")
+				epClient = NewSecondaryEndpointClient(nl, plc, ep)
 			} else {
-				logger.Info("OVS client")
-				if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
-					nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
-				}
-
-				epClient = NewOVSEndpointClient(
-					nw,
-					epInfo,
-					hostIfName,
-					contIfName,
-					vlanid,
-					localIP,
-					nl,
-					ovsctl.NewOvsctl(),
-					plc)
+				logger.Info("Transparent client")
+				epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 			}
-		} else if nw.Mode != opModeTransparent {
-			logger.Info("Bridge client")
-			epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
-		} else {
-			logger.Info("Transparent client")
-			epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 		}
-	}
 
-	// Cleanup on failure.
-	defer func() {
-		if err != nil {
-			logger.Error("CNI error. Delete Endpoint and rules that are created", zap.Error(err), zap.String("contIfName", contIfName))
-			endpt := &endpoint{
-				Id:                       epInfo.Id,
-				IfName:                   contIfName,
-				HostIfName:               hostIfName,
-				LocalIP:                  localIP,
-				IPAddresses:              epInfo.IPAddresses,
-				DNS:                      epInfo.DNS,
-				VlanID:                   vlanid,
-				EnableSnatOnHost:         epInfo.EnableSnatOnHost,
-				EnableMultitenancy:       epInfo.EnableMultiTenancy,
-				AllowInboundFromHostToNC: epInfo.AllowInboundFromHostToNC,
-				AllowInboundFromNCToHost: epInfo.AllowInboundFromNCToHost,
+		//nolint:gocritic
+		defer func(client EndpointClient, contIfName string) {
+			// Cleanup on failure.
+			if err != nil {
+				logger.Error("CNI error. Delete Endpoint and rules that are created", zap.Error(err), zap.String("contIfName", contIfName))
+				if containerIf != nil {
+					client.DeleteEndpointRules(ep)
+				}
+				// set deleteHostVeth to true to cleanup host veth interface if created
+				//nolint:errcheck // ignore error
```

Comment:
do you have UT for secondary endpoint failure case and see its cleaning up all the things it set up

Reply:
there shouldn't be anything to cleanup in secondary endpoints at the moment

Reply:
CNI should.. it cannot rely on containerd.. on delete cni should move interface back to VM without ip being assigned

Reply:
it looked like the secondary nic resets after moving between VM and pod namespace though

Reply:
moving interface back to default ns on delete now